packages/datetime/src/dateRangeInput.tsx

```diff
@@ -190,8 +221,114 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
         e.stopPropagation();
     }
 
-    private handleInputFocus = () => {
-        this.setState({ isOpen: true });
+    private handleStartInputFocus = (e: React.FormEvent<HTMLInputElement>) => {
+        this.handleInputFocus(e, DateRangeBoundary.START);
+    }
+
+    private handleEndInputFocus = (e: React.FormEvent<HTMLInputElement>) => {
+        this.handleInputFocus(e, DateRangeBoundary.END);
+    }
+
+    private handleInputFocus = (_e: React.FormEvent<HTMLInputElement>, boundary: DateRangeBoundary) => {
+        const { keys, values } = this.getStateKeysAndValuesForBoundary(boundary);
+        const inputString = this.getFormattedDateString(values.selectedValue);
+
+        this.setState({
+            isOpen: true,
+            [keys.inputString]: inputString,
+            [keys.isInputFocused]: true,
+        });
+    }
+
+    private handleStartInputBlur = (e: React.FormEvent<HTMLInputElement>) => {
+        this.handleInputBlur(e, DateRangeBoundary.START);
+    }
+
+    private handleEndInputBlur = (e: React.FormEvent<HTMLInputElement>) => {
+        this.handleInputBlur(e, DateRangeBoundary.END);
+    }
+
+    private handleInputBlur = (_e: React.FormEvent<HTMLInputElement>, boundary: DateRangeBoundary) => {
+        const { keys, values } = this.getStateKeysAndValuesForBoundary(boundary);
+
+        const maybeNextValue = this.dateStringToMoment(values.inputString);
+        const isValueControlled = this.props.value !== undefined;
```

Comment:
Yeah, I thought about that too. We should just be careful to note that `null` may be a meaningful value for `this.props.value`, so not every `isControlled()` implementation will be the same (some will need a strict `!== undefined` check).